### ISSUE TITLE
stack setup in CentOS 7 installs GHC with missing `libtinfo` versions.

### ISSUE THREAD

**Issue by stschiff** [commercialhaskell/stack#5561]
### General summary/comments

When installing stack using `curl -sSL https://get.haskellstack.org/ | sh` on CentOS 7, followed by a generic `stack setup` outside any project, while installing GHC it complains about version infos in the `libtinfo` libraries.

### Steps to reproduce

Here is a minimal `Dockerfile` for reproduction:
```
FROM centos:7

RUN yum -y update
RUN yum -y install zlib-devel wget ncurses-devel ncurses-compat-libs 
RUN curl -sSL https://get.haskellstack.org/ | sh
```

We can then build this Docker image via `docker build -t linux -f Dockerfile`
and then start a shell inside with `docker run -it linux /bin/bash`. 

When we then run `stack setup`, we get:
```
Writing implicit global project config file to: /root/.stack/global-project/stack.yaml
Note: You can change the snapshot via the resolver field there.
Using latest snapshot resolver: lts-17.14
Preparing to install GHC to an isolated location.
This will not interfere with any system-level installation.
Downloaded ghc-8.10.4.                                      
/root/.stack/programs/x86_64-linux/ghc-8.10.4.temp/ghc-8.10.4/configure: line 5957:   419 Aborted                 ./conftest
/root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/ghc-pkg: /lib64/libtinfo.so.5: no version information available (required by /root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/ghc-pkg)
/root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/ghc-pkg: /lib64/libtinfo.so.5: no version information available (required by /root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/../terminfo-0.4.1.4/libHSterminfo-0.4.1.4-ghc8.10.4.so)
/root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/ghc-pkg: /lib64/libtinfo.so.5: no version information available (required by /root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/ghc-pkg)
/root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/ghc-pkg: /lib64/libtinfo.so.5: no version information available (required by /root/.stack/programs/x86_64-linux/ghc-8.10.4/lib/ghc-8.10.4/bin/../terminfo-0.4.1.4/libHSterminfo-0.4.1.4-ghc8.10.4.so)
Installed GHC.    
stack will use a sandboxed GHC it installed
For more information on paths, see 'stack path' and 'stack exec env'
To use this GHC and packages outside of a project, consider using:
stack ghc, stack ghci, stack runghc, or stack exec
```

It turns out that this faulty GHC install also creates problems when trying to actually use it to install certain libraries (e.g. `cryptonite`.

### Expected
No complaint when installing GHC

### Actual
See output above.

### Stack version

```
[root@84509ded2382 /]# stack --version
Version 2.7.1, Git revision 8afe0c2932716b0441cf4440d6942c59568b6b19 x86_64 hpack-0.34.4
```

### Method of installation
See above `curl -sSL https://get.haskellstack.org/ | sh`

**Comment by hasufell** [comment]
Stack appears to download https://downloads.haskell.org/~ghc/8.10.6/ghc-8.10.6-x86_64-deb9-linux.tar.xz instead of https://downloads.haskell.org/~ghc/8.10.6/ghc-8.10.6-x86_64-centos7-linux.tar.xz

```
2021-09-02 22:44:21.149127: [debug] Asking for a supported GHC version
2021-09-02 22:44:21.149218: [debug] Installed tools:
 -
2021-09-02 22:44:21.149298: [debug] Run process: /sbin/ldconfig -p
2021-09-02 22:44:21.149813: [debug] Process finished in 0ms: /sbin/ldconfig -p
2021-09-02 22:44:21.149886: [debug] Found shared library libtinfo.so.5 in 'ldconfig -p' output
2021-09-02 22:44:21.149973: [debug] Did not find shared library libtinfo.so.6
2021-09-02 22:44:21.150015: [debug] Did not find shared library libncursesw.so.6
2021-09-02 22:44:21.150481: [debug] Found shared library libgmp.so.10 in 'ldconfig -p' output
2021-09-02 22:44:21.150529: [debug] Did not find shared library libgmp.so.3
2021-09-02 22:44:21.150560: [debug] Potential GHC builds: standard
2021-09-02 22:44:21.150601: [debug] Found already installed GHC builds:
2021-09-02 22:44:21.586134: [debug] Trying to setup GHC build: standard
2021-09-02 22:44:21.586268: [info] Preparing to install GHC to an isolated location.
2021-09-02 22:44:21.586325: [info] This will not interfere with any system-level installation.
2021-09-02 22:44:21.587617: [debug] Downloading from http://downloads.haskell.org/~ghc/8.10.6/ghc-8.10.6-x86_64-deb9-linux.tar.xz to /root/.stack/programs/x86_64-linux/ghc-8.10.6.tar.xz ...
2021-09-02 22:44:21.587707: [debug] Will check against sha1 hash: daea0db60a526a786f23ebd514d5aa110b99dd1e
2021-09-02 22:44:21.587769: [debug] Will check against sha256 hash: c14b631437ebc867f1fe1648579bc1dbe1a9b9ad31d7c801c3c77639523a83ae
2021-09-02 22:44:21.587850: [debug] Downloading /~ghc/8.10.6/ghc-8.10.6-x86_64-deb9-linux.tar.xz
```

Stack doesn't map OS to bindists, instead it discovers certain libraries etc., check out:

https://github.com/commercialhaskell/stack/blob/708b516a8d34cc42b6c533d161b18b31f913c3d6/src/Stack/Setup.hs#L915-L1012

What it discovers here is [linux64](https://github.com/commercialhaskell/stackage-content/blob/c61a95afb48d5db92fcedda82d0aed66b3c793ee/stack/stack-setup-2.yaml#L430) key, which pulls in deb9.

One way to work around it is to add this to `~/.stack/config.yaml`:

```yml
setup-info:
  ghc:
    linux64:
        8.10.6:
            url: "https://downloads.haskell.org/~ghc/8.10.6/ghc-8.10.6-x86_64-centos7-linux.tar.xz"
            content-length: 203488844
            sha1: df62ce78fa912c83fab3b4400617b965092f33f2
            sha256: 71bb69b7185cbb44a941f5214b241a55e158f69b103f10214a05de7845f77e45
```

----

In order to move forward, we need to figure out what exactly is the difference between the deb9 and centos7 bindists.

**Comment by stschiff** [comment]
Thanks for that thorough investigation! Very interesting that it simply downloads a wrong bindist. I'll take a look at the centos7-version. I'm not too familiar with the internals of stack, but the fix you're proposing may already solve the issue at focus for me... but of course it'd be more preferable to solve it more generally.

**Comment by etherz10** [comment]
Been looking for a fix for this for a while now. Much appreciated!

Would be great if this could be resolved without needing to remember to hack the stack config on each machine.

**Comment by etherz10** [comment]
@mpilgrem Are you able to assist with this at all? I'm happy to test a fix whenever it becomes available

**Comment by mpilgrem** [comment]
@etherz10, I have tried to help (see below), but I don't have good news.

@hasufell has already described Stack's general approach above: Stack looks for the presence of certain key C libraries and then picks a version of GHC built against certain libraries. The most 'out-of-date' (if I may use that term) Linux distribution that Stack currently supports 'out of the box' is Debian 9 'stretch'.  Debian 9 'stretch' comes with `libtinfo5` (version 6.0), `libncurses5` (version 6.0), `'libgmp10` (version 6.1.2) and `libc6` (version 2.24).

The first question is, what key C libraries come with CentOS 7? I looked at this list: http://mirror.centos.org/centos-7/7/os/x86_64/Packages/, and spotted:
* no references to `tinfo`
* `ncurses-5.9-14.20130511.el7_4.x86_64.rpm` I assume that is version 5.9 of ncurses
* `gmp-6.0.0-15.el7.x86_64.rpm` I assume that is version 6.0.0 of the GNU Multiple Precision Arithmetic Library
* `glibc-2.17-317.el7.x86_64.rpm` I assume that is version 2.17 of the GNU C Library

So, the 'problem' appears to be that CentOS 7 comes with key C libraries that are more 'out-of-date' than those that come with Debian 9. That is why Stack 2.7.1 was defaulting to the 'Debian 9' version of GHC. Stack 2.9.3 will do the same.

I think there is no prospect of Stack extending its backwards compatibility 'out of the box', especially as I [read here](https://www.zdnet.com/article/centos-linux-8-is-about-to-die-what-do-you-do-next/) that CentOS Linux was discontinued at the end of 2021.

As such, CentOS 7 will have to be treated as something that needs a custom `setup-info` directory entry in Stack's non-project specific configuration: https://docs.haskellstack.org/en/stable/yaml_configuration/#setup-info.
